### PR TITLE
Remove all refs to enkf_main obj in ERT

### PR DESCRIPTION
**Pull request by oysteoh**
With this PR i've tried loosen up on the strong dependency towards enkf_main making full use of the enkf_facade interface. Instead of passing around the global ERT.ert (enkf_main) object, i've just used the global one. The enkf_facade is still leaking some objects out into ERT unfortunately, but my hope is that this is a good start and will make it easier moving implementation details into the facade as we go... This PR is big enough i feel and thus do not want to start refactoring more logic and move to the facade ( which probably should be done in the future )

Benefit of having all enkf_main code and implementation details in a facade will make our life easier as we go on with refactoring of the enkf_main implementation. In ERT we'll just need to make adjustments one place.

### CODE REVIEW

File: ert_gui/simulation/run_dialog.py [hunk 1]
```diff
@@ -90,7 +86,7 @@ def __init__(self, run_model, parent):
         self.plot_tool.setParent(None)
         self.plot_button = QPushButton(self.plot_tool.getName())
         self.plot_button.clicked.connect(self.plot_tool.trigger)
-        self.plot_button.setEnabled(ert is not None)
+        self.plot_button.setEnabled(True)
```
Comment (sondreso) [inline ert_gui/simulation/run_dialog.py]: Is there no chance that `ert = run_model.ert()` returns None?

File: ert_gui/simulation/run_dialog.py [hunk 2]
```diff
@@ -90,7 +86,7 @@ def __init__(self, run_model, parent):
         self.plot_tool.setParent(None)
         self.plot_button = QPushButton(self.plot_tool.getName())
         self.plot_button.clicked.connect(self.plot_tool.trigger)
-        self.plot_button.setEnabled(ert is not None)
```
Comment (sondreso) [inline ert_gui/simulation/run_dialog.py]: Is there no chance that ert = run_model.ert() returns None?

File: ert_gui/tools/plot/data_type_keys_list_model.py [hunk 3]
```diff
@@ -42,14 +34,14 @@ def data(self, index, role=None):
...
-            items = self.keyManager().allDataTypeKeys()
+            items = ERT.enkf_facade.get_all_data_type_keys()
             row = index.row()
             item = items[row]
 
             if role == Qt.DisplayRole:
                 return item
             elif role == Qt.BackgroundRole:
-                if self.keyManager().isKeyWithObservations(item):
+                if ERT.enkf_facade.is_key_with_observations(item):
```
Comment (sondreso) [inline ert_gui/tools/plot/data_type_keys_list_model.py]: Extra spaces?

**Comment by markusdregi** [comment]
This is a good idea! I have a couple of initial comments before a more thorough review is conducted...

- Should we consider calling it res_facade, as enkf is a rather outdated name...?
- I’m no big fan of Singletons and/or massive global state. When you are removing ert as a function argument, you are enforcing this pattern. Would it be an option to pass in the res_facade?

File: ert_gui/simulation/run_dialog.py (hunk 2 above)
Comment (oysteoh) [reply ert_gui/simulation/run_dialog.py]: Actually I'm not sure - but from my understanding it's currently impossible for ert to be None - but i may be mistaken here... previously an assertion would make assertion error if run_model is not a BaseRunModel, the same check is used in if where ert was set... as far as i understand the if could never fail previously. This check is removed - the whole ert assignment is removed - and now the global ERT is used where ert has to be set when initializing the code - either through GUI or CLI

```
        assert isinstance(run_model, BaseRunModel)
        self._run_model = run_model

        ert = None
        if isinstance(run_model, BaseRunModel):
            ert = run_model.ert()
```

**Comment by oysteoh** [comment]
> This is a good idea! I have a couple of initial comments before a more thorough review is conducted...
> 
>     * Should we consider calling it res_facade, as enkf is a rather outdated name...?
> 
>     * I’m no big fan of Singletons and/or massive global state. When you are removing ert as a 
function argument, you are enforcing this pattern. Would it be an option to pass in the res_facade?

- res_facade is a better name - agreed! 
- yes - i'm a bigger fan of injecting interfaces rather than using singletons. the reason i ended up with a singleton while holding my nose was that the global ERT is much more used and it would cause bigger changes introducing dependency injection. But - if the reviewers want's DI i would be glad to introduce the concept instead of singleton :)

**Comment by markusdregi** [comment]
> But - if the reviewers want's DI i would be glad to introduce the concept instead of singleton :)

How about if we passed `res_facade` to the methods that used to get `ert` to avoid further contamination of the _singleton_ and then we create a separate issue of removing the _singleton_ completely?

**Comment by oysteoh** [comment]
> > But - if the reviewers want's DI i would be glad to introduce the concept instead of singleton :)
> 
> How about if we passed `res_facade` to the methods that used to get `ert` to avoid further contamination of the _singleton_ and then we create a separate issue of removing the _singleton_ completely?

I could do that - probably nice doing it in smallest steps possible. I will rework this PR - inject res_facade instead of enkf_main obj - and - do some small refactoring where ert has been exposed as variable ( like _ert = ERT.ert ) . We'll see how it pans out :+1:

**Comment by andreabrambilla** [comment]
@oysteoh mentioned this can be closed as there is more work to be done here

**Pull request closed without merging**